### PR TITLE
Feature/rmk 29 refactor internal logic of rmk cluster provider initialization and selection

### PR DESCRIPTION
**Pull request by apanasiuk-el**

### CODE REVIEW

File: cmd/project.go [hunk 1]
```diff
@@ -67,9 +67,61 @@ func newProjectCommand(conf *config.Config, ctx *cli.Context, workDir string) *P
...
+
+	if p.Ctx.IsSet("environments") {
+		p.projectFile.Spec.Environments = make(map[string]*config.ProjectRootDomain)
+		for _, val := range p.Ctx.StringSlice("environments") {
+			if len(val) > 0 {
+				matchRootDomain := regexp.MustCompile(`^.+\.root-domain=.+$`).MatchString(val)
+				splitRootDomain := strings.SplitN(val, ".", 2)
+
+				if !matchRootDomain && len(splitRootDomain) == 2 {
+					return fmt.Errorf("option %s for environment %s set not correctly",
```
Comment (anovikov-el) [inline cmd/project.go:86]: `"option %s for environment %s not set correctly"`

File: cmd/flags.go [hunk 2]
```diff
@@ -182,6 +152,21 @@ func flagsProjectGenerate() []cli.Flag {
 			Usage:   "create SOPS age keys for generated project structure",
 			Aliases: []string{"c"},
 		},
+		&cli.StringSliceFlag{
```
Comment (anovikov-el) [inline cmd/flags.go:155]: suggest using `list` instead of `list of` because these are all verbs.

**Review by anovikov-el — CHANGES_REQUESTED** [review]

**Review by anovikov-el — DISMISSED** [review]

File: cmd/cluster.go [hunk 3]
```diff
@@ -1,55 +1,412 @@
...
+
+	_, currentContextName, err := cc.getKubeContext()
+	if err != nil {
+		return "", "", err
+	}
+
+	cc.SpecCMD = cc.prepareHelmfile("--log-level", "error", "-l", "config=clusterctl", "template")
+	cc.SpecCMD.DisableStdOut = true
+	if err := releaseRunner(cc).runCMD(); err != nil {
+		return "", "", fmt.Errorf("Helmfile failed to render template by label release: config=clusterctl\n%s",
```
Comment (anovikov-el) [inline cmd/cluster.go:102]: should be `... by release label: ...`

File: cmd/config.go [hunk 4]
```diff
@@ -531,7 +538,56 @@ func configInitAction(conf *config.Config, gitSpec *git_handler.GitSpec) cli.Act
...
+			if util.IsExists(
+				util.GetHomePath(azure_provider.AzureHomeDir, azure_provider.AzurePrefix+gitSpec.ID+".json"), true) {
+				if err := ac.ReadSPCredentials(gitSpec.ID); err != nil {
+					return err
+				}
+			}
+
+			if c.Bool("azure-service-principle") {
+				if err := json.NewDecoder(os.Stdin).Decode(&asp); err != nil {
+					return fmt.Errorf("unable to deserialize json from stdin: %s", err.Error())
```
Comment (anovikov-el) [inline cmd/config.go:555]: maybe `json` should be in uppercase?
e.g. `JSON`

File: providers/azure_provider/azure.go [hunk 5]
```diff
@@ -1 +1,153 @@
...
+	ac.ClientID = asp.AppId
+	ac.ClientSecret = asp.Password
+	ac.TenantID = asp.Tenant
+}
+
+func getTagStructName(i interface{}, name string) error {
+	if field, ok := reflect.TypeOf(i).Elem().FieldByName(name); ok {
+		return fmt.Errorf("service principal option %s not defined", strings.TrimSuffix(field.Tag.Get("json"), ",omitempty"))
+	} else {
+		return fmt.Errorf("field by name %s not defined", name)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:61]: `field with name %s not defined`

File: cmd/cluster.go (hunk 3 above)
Comment (apanasiuk-el) [reply cmd/cluster.go:102]: fixed

File: cmd/config.go (hunk 4 above)
Comment (apanasiuk-el) [reply cmd/config.go:555]: fixed

File: providers/azure_provider/azure.go (hunk 5 above)
Comment (apanasiuk-el) [reply providers/azure_provider/azure.go:61]: fixed

File: cmd/flags.go (hunk 2 above)
Comment (apanasiuk-el) [reply cmd/flags.go:188]: fixed

File: cmd/project.go (hunk 1 above)
Comment (apanasiuk-el) [reply cmd/project.go]: fixed

**Review by anovikov-el — DISMISSED** [review]

File: cmd/flags.go [hunk 6]
```diff
@@ -63,6 +63,13 @@ func flagsConfig() []cli.Flag {
 			Aliases:  []string{"azp"},
 			EnvVars:  []string{"RMK_AZURE_CLIENT_SECRET", "AZURE_CLIENT_SECRET"},
 		},
+		&cli.StringFlag{
+			Category: azureFlagsCategory,
+			Name:     "azure-location",
+			Usage:    "Azure location region",
```
Comment (anovikov-el) [inline cmd/flags.go:69]: should be just "Azure location"?

File: providers/azure_provider/azure.go [hunk 7]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+func (ac *AzureConfigure) createKeyVaultResourceGroup(tenant string) error {
+	params := armresources.ResourceGroup{
+		Location: to.Ptr(ac.Location),
+	}
+
+	update, err := ac.GroupClient.CreateOrUpdate(ac.Ctx, tenant+"-"+util.SopsRootName, params, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to create Azure resource group: %s", tenant+"-"+util.SopsRootName)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:198]: should be "permission denied..."

File: providers/azure_provider/azure.go [hunk 8]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+
+	return nil
+}
+
+func (ac *AzureConfigure) existsKeyVaultResourceGroup(tenant string) (bool, error) {
+	existence, err := ac.GroupClient.CheckExistence(ac.Ctx, tenant+"-"+util.SopsRootName, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to check existence Azure resource group: %s",
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:217]: should be "permission denied..."

Comment (anovikov-el) [reply providers/azure_provider/azure.go:217]: "... existence **of** Azure resource group: ..."

File: providers/azure_provider/azure.go [hunk 9]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+			},
+			TenantID: to.Ptr(ac.TenantID),
+		},
+	}
+
+	update, err := ac.VaultsClient.BeginCreateOrUpdate(ac.Ctx, ac.ResourceGroupName, ac.KeyVaultName, params, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to create Azure key vault: %s",
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:255]: same

File: providers/azure_provider/azure.go [hunk 10]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+
+	resp, err := ac.VaultsClient.Get(ac.Ctx, tenant+"-"+util.SopsRootName, ac.KeyVaultName, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 404 {
+			return false, nil
+		}
+
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to get Azure Key Vault: %s", ac.KeyVaultName)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:293]: seems that we should fixed the naming in every log line to be: "Azure **K**ey **V**ault"
some places have: "created Azure **k**ey **v**ault"

File: providers/azure_provider/azure.go [hunk 11]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+
+	params := azsecrets.SetSecretParameters{
+		Value: to.Ptr(value),
+	}
+
+	secret, err := client.SetSecret(ac.Ctx, keyName, params, nil)
+	if err != nil {
+		var respErr *azcore.ResponseError
+		if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+			zap.S().Warnf("permissions denied to create Azure key vault secret: %s", keyName)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:319]: same

File: providers/azure_provider/azure.go [hunk 12]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+			return nil
+		}
+
+		return err
+	}
+
+	zap.S().Infof("created Azure key vault secret: %s, %s", keyName, *secret.ID)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:326]: same

File: providers/azure_provider/azure.go [hunk 13]
```diff
@@ -153,17 +186,191 @@ func (ac *AzureConfigure) GetAzureClusterContext(groupName, clusterName string)
...
+
+		for _, val := range page.Value {
+			name := val.ID.Name()
+			version := val.ID.Version()
+
+			secret, err := client.GetSecret(ac.Ctx, name, version, nil)
+			if err != nil {
+				var respErr *azcore.ResponseError
+				if errors.As(err, &respErr) && respErr.StatusCode == 403 {
+					zap.S().Warnf("permissions denied to get of Azure key vault secret: %s", name)
```
Comment (anovikov-el) [inline providers/azure_provider/azure.go:364]: same

**Review by anovikov-el — CHANGES_REQUESTED** [review]

File: providers/azure_provider/azure.go (hunk 10 above)
Comment (apanasiuk-el) [reply providers/azure_provider/azure.go:293]: key vault is not a common noun

**Review by anovikov-el — DISMISSED** [review]

File: providers/google_provider/gcp.go [hunk 14]
```diff
@@ -58,40 +78,195 @@ func (gcp *GCPConfigure) CopySACredentials(fileSuffix string) error {
...
+				return nil, err
+			}
+		}
+
+		accReq := &secretmanagerpb.AccessSecretVersionRequest{Name: resp.Name + "/versions/latest"}
+		result, err := client.AccessSecretVersion(gcp.Ctx, accReq)
+		if err != nil {
+			var respError *apierror.APIError
+			if errors.As(err, &respError) && respError.GRPCStatus().Code().String() == gRPCErrorPermissionDenied {
+				zap.S().Warnf("permission denied to get access for GCP secrets values")
```
Comment (anovikov-el) [inline providers/google_provider/gcp.go:128]: permission denied to get access **to** GCP secrets values

**Review by vadim-bogdanovsky — APPROVED** [review]

File: providers/google_provider/gcp.go [hunk 15]
```diff
@@ -58,40 +78,195 @@ func (gcp *GCPConfigure) CopySACredentials(fileSuffix string) error {
...
+	secretVersionReq := &secretmanagerpb.AddSecretVersionRequest{
+		Parent:  fmt.Sprintf("projects/%s/secrets/%s", gcp.ProjectID, keyName),
+		Payload: &secretmanagerpb.SecretPayload{Data: value},
+	}
+
+	version, err := client.AddSecretVersion(gcp.Ctx, secretVersionReq)
+	if err != nil {
+		var respError *apierror.APIError
+		if errors.As(err, &respError) && respError.GRPCStatus().Code().String() == gRPCErrorPermissionDenied {
+			zap.S().Warnf("permission denied to add to secret %s value", keyName)
```
Comment (anovikov-el) [inline providers/google_provider/gcp.go:193]: permission denied to add GCP secret %s value

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by anovikov-el — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by anovikov-el — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Review by anovikov-el — DISMISSED** [review]

**Review by vadim-bogdanovsky — APPROVED** [review]

**Pull request merged**